packages/editor/src/editor.ts

```diff
@@ -0,0 +1,207 @@
+    }
+
+    try {
+      const extensions = await fs.readdir(path.join(this._vscodeDir, 'extensions'));
+      const hasMongodbExtension = !!extensions
+        .find((name) => name.includes('mongodb.mongodb-vscode'));
+
+      this.messageBus.emit('mongosh-editor:read-vscode-extensions-done', {
+        vscodeDir: this._vscodeDir,
+        hasMongodbExtension
+      });
+
+      return hasMongodbExtension ? 'mongodb' : 'js';
+    } catch (error) {
+      this.messageBus.emit('mongosh-editor:read-vscode-extensions-failed', {
+        action: 'mongosh-editor:read-vscode-extensions-failed',
+        error: (error as Error).message
+      });
+
+      return 'js';
+    }
+  }
+
+  async _getEditor(): Promise<string|null> {
+    // Check for an external editor in the mongosh configuration.
+    let editor: string | null = await this._instanceState.shellApi.config.get('editor');
+
+    // Check for an external editor in the environment variable.
+    if (!editor && process.env.EDITOR) {
+      editor = process.env.EDITOR;
+    }
+
+    return editor;
+  }
+
+  async _createTempFile({ content, ext }: { content: string; ext: string }): Promise<string> {
+    const tmpDoc = path.join(this._tmpDir, `edit-${new bson.ObjectId()}.${ext}`);
+
+    // Create a temp file to store content that is being edited.
+    await fs.mkdir(path.dirname(tmpDoc), { recursive: true, mode: 0o700 });
+    await fs.writeFile(tmpDoc, beautify(content));
+
+    return tmpDoc;
+  }
+
+  async _readTempFile(tmpDoc: string): Promise<string> {
+    const content = await fs.readFile(tmpDoc, 'utf8');
+    await fs.unlink(tmpDoc);
+    return this._makeMultilineJSIntoSingleLine(content);
+  }
+
+  _isVscodeApp(cmd: string): boolean {
+    const regex = /^(.*)[\/\\]?[cC]ode(.exe)?$/;
+    return regex.test(cmd);
+  }
+
+  _isIdentifier(code: string): boolean {
+    const regex = /^([^!#%&()*+,\-/\\^`{|}~]+)$/;
+    return regex.test(code);
+  }
+
+  async _getEditorContent(code: string): Promise<string> {
+    if (!code) {
+      return this._content;
+    }
+
+    if (!this._isIdentifier(code)) {
+      return code;
+    }
+
+    const evalResult = await this._loadExternalCode(code, '@(shell eval)');
+    return evalResult.toString();
+  }
+
+  async runEditCommand(args: string[]): Promise<void> {
+    await this.print('Opening an editor...');
+
+    const code = args.join(' ');
+    const editor: string|null = await this._getEditor();
+
+    // If none of the above configurations are found return an error.
+    if (!editor) {
+      throw new Error('Command failed with an error: please define an external editor');
+    }
+
+    const content = await this._getEditorContent(code);
+    const ext = await this._getExtension(editor);
+    const tmpDoc = await this._createTempFile({ content, ext });
+
+    this.messageBus.emit('mongosh-editor:run-edit-command', {
+      tmpDoc,
+      editor,
+      code
+    });
+
+    const proc = spawn(editor, [tmpDoc], { stdio: 'inherit' });
```

Comment:
I think this is definitely better than the previous state, but now there’s the reverse problem, because I guess it could still be reasonable for somebody to specify something in `editor` that *does* contain multiple parts (for example, I use `EDITOR='nano -cA'` locally). We can defer that to another subtask, but I think we should definitely still address it.

I’m pretty sure that we’ll need to spawn the editor using a shell by default – there is an option for that in the Node.js child_process for sure, maybe that’s even enough, maybe not.

(So we should support:

1.`EDITOR=code`
2. `EDITOR=C:\Program Files\Editor\editor.exe`
3. `EDITOR=nano -cA`

and since we *cannot* easily distinguish between the semantics of the spaces between cases 2 and 3, we will need the OS shell to do that for us.)

Reply:
Created a ticket for that: https://jira.mongodb.org/browse/MONGOSH-996